### PR TITLE
Feature aws s3 upload

### PR DESCRIPTION
**Pull request by clothe09986**

### CODE REVIEW

**Comment by clothe09986** [comment]
[【影片】媒體庫、前台串接aws影片](https://trello.com/c/wXRAVpZT/4477-%E3%80%90%E5%BD%B1%E7%89%87%E3%80%91%E5%AA%92%E9%AB%94%E5%BA%AB%E3%80%81%E5%89%8D%E5%8F%B0%E4%B8%B2%E6%8E%A5aws%E5%BD%B1%E7%89%87)

File: src/utility/storage/storage.controller.ts [hunk 1]
```diff
@@ -2,21 +2,76 @@ import { Body, Controller, Post } from '@nestjs/common';
...
+      };
+      file: {
+        name: string;
+        type: string;
+        size: number;
+      };
+      appId: string;
+      authorId: string;
+      attachmentId: string;
+    },
```
Comment (jackey8616) [inline src/utility/storage/storage.controller.ts]: Shall write in DTO.

File: src/utility/storage/storage.service.ts [hunk 2]
```diff
@@ -81,6 +94,40 @@ export class StorageService {
...
+      id: attachmentId,
+      appId,
+      author: { id: authorId },
+      name: file.name,
+      filename: file.name,
+      contentType: file.type,
+      size: file.size,
+      options: [{ source: 's3' }],
+    });
+  }
```
Comment (jackey8616) [inline src/utility/storage/storage.service.ts]: Method `insertVideoAttachment` shall write in MediaInfra & wrap in @ MediaService.  
MediaService should import @ `StorageController`'s constructor and invoke @ there.

File: src/utility/storage/storage.controller.ts [hunk 3]
```diff
@@ -2,21 +2,76 @@ import { Body, Controller, Post } from '@nestjs/common';
...
+  ) {
+    const { params, file, appId, authorId, attachmentId } = body;
+    const result = await this.storageService.completeMultipartUpload(params);
+    await this.storageService.insertVideoAttachment(appId, authorId, attachmentId, file);
```
Comment (jackey8616) [inline src/utility/storage/storage.controller.ts]: `insertVideoAttachment ` should not write in `storageService`, it should be `mediaService`.

**Review by jackey8616 — CHANGES_REQUESTED** [review]

File: src/utility/storage/storage.service.ts (hunk 2 above)
Comment (clothe09986) [reply src/utility/storage/storage.service.ts]: fixed

File: src/utility/storage/storage.controller.ts (hunk 3 above)
Comment (clothe09986) [reply src/utility/storage/storage.controller.ts]: fixed

File: src/media/media.service.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+export class MediaService {
+  private readonly awsCloudfrontKeyPairId: string;
+  private readonly awsCloudfrontPrivateKey: string;
+
+  constructor(
+    @InjectEntityManager() private readonly entityManager: EntityManager,
+    private readonly mediaInfra: MediaInfrastructure,
+    private readonly configService: ConfigService<{
+      AWS_CLOUDFRONT_KEY_PAIR_ID: string;
+      AWS_CLOUDFRONT_PRIVATE_KEY: string;
```
Comment (jackey8616) [inline src/media/media.service.ts]: These keys is not add into `.env` & `.env.example`  .

File: src/utility/storage/storage.controller.ts [hunk 5]
```diff
@@ -2,21 +2,79 @@ import { Body, Controller, Post } from '@nestjs/common';
...
+      };
+      appId: string;
+      authorId: string;
+      attachmentId: string;
+    },
+  ) {
+    const { params, file, appId, authorId, attachmentId } = body;
+    const result = await this.storageService.completeMultipartUpload(params);
+    await this.mediaService.insertAttachment(appId, authorId, attachmentId, file.name, file.type, file.size, {
+      source: 's3',
```
Comment (jackey8616) [inline src/utility/storage/storage.controller.ts]:
```ts
{
  source: 's3,
}
```

or 

```ts
{
  source: {
    s3: 's3://.....',
  },
}
```
?

File: src/utility/storage/storage.service.ts [hunk 6]
```diff
@@ -10,25 +10,34 @@ import {
...
 export class StorageService {
   private readonly awsS3RegionStorage: string;
+  private readonly awsS3RegionStatic: string;
   private readonly awsS3BucketStatic: string;
   private readonly awsS3BucketStorage: string;
 
   constructor(
     private readonly configService: ConfigService<{
       AWS_S3_BUCKET_STATIC: string;
+      AWS_S3_REGION_STATIC: string;
```
Comment (jackey8616) [inline src/utility/storage/storage.service.ts]: This key is not add into `.env` & `.env.example`  .

File: src/utility/storage/storage.controller.ts [hunk 7]
```diff
@@ -2,21 +2,79 @@ import { Body, Controller, Post } from '@nestjs/common';
...
+    @Body()
+    body: {
+      params: {
+        Key: string;
+        ContentType: string;
+      };
+    },
+  ) {
+    const { params } = body;
+    const { UploadId } = await this.storageService.createMultipartUpload(params);
```
Comment (jackey8616) [inline src/utility/storage/storage.controller.ts]: Exposing too many parameters.  
Please wrap `createMultipartUpload` method with limited parameters.

File: src/utility/storage/storage.controller.ts [hunk 8]
```diff
@@ -2,21 +2,79 @@ import { Body, Controller, Post } from '@nestjs/common';
...
+    body: {
+      params: {
+        Key: string;
+        UploadId: string;
+        PartNumber: number;
+      };
+    },
+  ) {
+    const { params } = body;
+    const presignedUrl = await this.storageService.getSignedUrlForUploadPartStorage(params, 60);
```
Comment (jackey8616) [inline src/utility/storage/storage.controller.ts]: Exposing too many parameters.  
Please wrap `getSignedUrlForUploadPartStorage ` method with limited parameters.

File: src/utility/storage/storage.controller.ts [hunk 9]
```diff
@@ -2,21 +2,79 @@ import { Body, Controller, Post } from '@nestjs/common';
...
+        type: string;
+        size: number;
+      };
+      appId: string;
+      authorId: string;
+      attachmentId: string;
+    },
+  ) {
+    const { params, file, appId, authorId, attachmentId } = body;
+    const result = await this.storageService.completeMultipartUpload(params);
```
Comment (jackey8616) [inline src/utility/storage/storage.controller.ts]: Exposing too many parameters.  
Please wrap `completeMultipartUpload ` method with limited parameters.

File: src/media/media.controller.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+export class MediaController {
+  constructor(private readonly mediaService: MediaService) {}
+
+  @Post('/sign-cookies') signCookie(@Res() res: Response, @Body() body: { url: string }) {
+    const cookies = this.mediaService.signCookies(body.url);
+    res.cookie('CloudFront-Expires', cookies['CloudFront-Expires']);
+    res.cookie('CloudFront-Key-Pair-Id', cookies['CloudFront-Key-Pair-Id']);
+    res.cookie('CloudFront-Signature', cookies['CloudFront-Signature']);
+    res.send('success set cookie');
+  }
```
Comment (jackey8616) [inline src/media/media.controller.ts]: Should `signedCookie` be inside `AuthService` & `AuthController` ?

**Review by jackey8616 — CHANGES_REQUESTED** [review]

File: src/media/media.controller.ts (hunk 10 above)
Comment (clothe09986) [reply src/media/media.controller.ts]: or `signedCookie` method is inside `AuthService` and `MediaController` invokes this method?

File: src/media/media.service.ts (hunk 4 above)
Comment (clothe09986) [reply src/media/media.service.ts]: fixed

File: src/utility/storage/storage.service.ts (hunk 6 above)
Comment (clothe09986) [reply src/utility/storage/storage.service.ts]: fixed

File: src/utility/storage/storage.controller.ts (hunk 1 above)
Comment (clothe09986) [reply src/utility/storage/storage.controller.ts]: fixed

File: src/utility/storage/storage.controller.ts (hunk 5 above)
Comment (clothe09986) [reply src/utility/storage/storage.controller.ts]: fixed

File: src/utility/storage/storage.controller.ts (hunk 7 above)
Comment (clothe09986) [reply src/utility/storage/storage.controller.ts]: fixed

File: src/utility/storage/storage.controller.ts (hunk 8 above)
Comment (clothe09986) [reply src/utility/storage/storage.controller.ts]: fixed

File: src/utility/storage/storage.controller.ts (hunk 9 above)
Comment (clothe09986) [reply src/utility/storage/storage.controller.ts]: fixed

File: src/media/media.controller.ts (hunk 10 above)
Comment (clothe09986) [reply src/media/media.controller.ts]: fixed

**Pull request closed without merging**